### PR TITLE
Spike devops

### PR DESCRIPTION
**Pull request by flyingoctopus**
Spike to add devops tooling to the base SimpleDJ project

- [x] containerized Linux build environment

Hopefuls:
- [x] CLion integration
- [ ] multiarch build targets (armv7 et al?)
- [ ] CI/CD beginnings
  - [ ] Profiler
  - [ ] Linter
- [ ] CLI tool
- [ ] Github runner integration
- [ ] migrate container image to alpine
- [ ] multiarch build

To set up the docker environment:
* make sure you have docker, docker-compose, and it's requirements installed on dev machine
* build container from project root using `docker build -t clion/ubuntu/cpp-env:1.0 -f Dockerfile .`
* go to Settings | Build, Execution, Deployment | Toolchains.
* click Add toolchain and select Docker
* click the gear button to the Server field to add a Docker image:
* select the Docker image and wait until the tools detection finishes.

Note: Use the Container Settings field to provide additional container settings, such as port and volume bindings:
Docker container settings

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Build and run:
+#   docker build -t clion/ubuntu/cpp-env:1.0 -f Dockerfile .
+#   docker run --rm -it clion/ubuntu/cpp-env:1.0 /bin/sh
+#   docker run -rm -it -v $(pwd):/app clion/ubuntu/cpp-env:1.0
```
Comment (filip-franek) [inline Dockerfile]: Probably do not need the 3rd line without the bind mount and want to keep the 4th line, right? Just add the missing hyphen there.  
```
# docker run --rm -it -v $(pwd):/app clion/ubuntu/cpp-env:1.0
```
I am a bit confused though whether I should build and run this image to be able to debug the code in CLion or whether I should be able to build and run the code in the container.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#TODO define target-specific env variables
+
+# set permissions
+RUN chmod 0755 -R /app
+
+# non interactive mode plays nice with CLion debug window
+RUN DEBIAN_FRONTEND="noninteractive" apt-get update && apt-get -y install tzdata
+
+# setup
+RUN apt-get update \
```
Comment (filip-franek) [inline Dockerfile:26]: Guess `apt-get update` is needed only once in a Dockerfile.

**Review by filip-franek — CHANGES_REQUESTED** [review]
Beware that I have never seen the repo, so my review might be affected by that.

Followed the notes in the PR and was able to build the app and run it, nice work!

image

I tried to mix some audio, but the app did not play it back ;(.

It would be good to add the procedure from the PR to `README.md` or `DEVELOPMENT.md` and emphasize running it in CLion. When I just looked at the folder for the first time and tried to build it with CMake, it built, but could not execute it.
```
cmake -B cmake-build-debug . && cmake --build cmake-build-debug

./cmake-build-debug/SimpleDJ_artefacts/SimpleDJ
malloc(): invalid size (unsorted)
Aborted (core dumped)
```

Also, you might want to add
- the info about building the image with `docker build -t clion/ubuntu/cpp-env:1.0 -f Dockerfile .` prior to setting Docker environment in CLion.
- setting Docker environment in CLion with `-v <SimpleDJ_repo>:/app --rm`.

File: Dockerfile (hunk 1 above)
Comment (flyingoctopus) [reply Dockerfile]: Hey Filip! So the last line is more for mounting a volume directly from the command line, which we don't have to do if we use clion in our dev environment. The second last line is to load an interactive shell for debug purposes. I will add more verbose comments to that line block.

File: Dockerfile (hunk 2 above)
Comment (flyingoctopus) [reply Dockerfile:26]: The main issue I ran into was that tzdata continued to require input despite the `-y` being appended to the install. The popular solution I found was to set the flag above to 'noninteractive' (example solution [here](https://serverfault.com/questions/949991/how-to-install-tzdata-on-a-ubuntu-docker-image)). I thought it'd be more readable to break that out atomically. I would be happy to merge that into the setup block.

Comment (filip-franek) [reply Dockerfile:26]: Thanks for pointing me to: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run @flyingoctopus  where I learned we should prepend each `apt-get install` with `apt-get update`. Further more, the `apt-get clean` does not seem to be necessary here and we can can further reduce the size of the image with `rm -rf /var/lib/apt/lists/*`.

**Review by filip-franek — APPROVED** [review]
As discussed, the PR looks good already. It would be good to only include the instruction on how to set up the dev environment as noted in your PR comment.

**Comment by flyingoctopus** [comment]
> As discussed, the PR looks good already. It would be good to only include the instruction on how to set up the dev environment as noted in your PR comment.

Sounds great @filip-franek. I will create a `DEVELOPMENT.md` file to the project root detailing setup instructions. There are some clion-specific steps I will have to review. Is this something you can validate in you dev environment? Do you have a CLion license @filip-franek ?

File: Dockerfile (hunk 1 above)
Comment (flyingoctopus) [reply Dockerfile]: typo fixed. resolving.

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+1. You should have docker installed in your system, if not click [here](https://docs.docker.com/get-docker/).
+
+2. Build the docker image:
+    ```shell
+    docker build -t clion/ubuntu/cpp-env:1.0 -f Dockerfile .
+    ```
+3. Set up CLion:
+    - Navigate to Settings > Build, Execution, Deployment > Toolchains
+    - Select _+_ then choose _Docker_ from the dropdown
+    - Click the gear icon on the right side of the server field
```
Comment (filip-franek) [inline DEVELOPMENT.md]: This step was not necessary in my setup.

image

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    - Select the docker image  from the dropdown field
+      labeled _Container Settings_ and wait for CLion to
+      complete the configuration.
```
Comment (filip-franek) [inline DEVELOPMENT.md]: what I did was:
- Select the docker image `clion/ubuntu/cpp-env:1.0` from the dropdown field 
(see screenshot above)
We might have difference CLion versions, thus the UI is different?

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    - Use the container setttings to mount the project root as a volume with the path `/tmp/SimpleDJ`
+		*NOTE*: It has to be an abolute path
+
+4. After configuring a Docker toolchain, you can select it in CMake profiles or in Makefile settings. Alternatively, move the toolchain to the top of the list to make it default.
```
Comment (filip-franek) [inline DEVELOPMENT.md]: It was successfully built, but could not run it or debug it. Getting this error `JUCE Assertion failure in juce_linux_XWindowSystem.cpp:1718`, but the README says, it should work on GNU/Linux systems with X-Windows which my machine fulfils + I was already able to run it in the past.

File: DEVELOPMENT.md (hunk 4 above)
Comment (flyingoctopus) [reply DEVELOPMENT.md]: Hey Filip, it looks like you already had a docker toolchain configuration set up on your copy of clion and that it's set up to use an already-built container called xtensa? My recommendation would be to create another docker-based Toolchain config  but rename is something unique to identify it from the previous config. I just reproduced the steps above and it should match. Let me know if that makes sense, I'd be happy to jump on a call. Cheers!

File: DEVELOPMENT.md (hunk 5 above)
Comment (flyingoctopus) [reply DEVELOPMENT.md]: I need to investigate this. It's unclear if this is related to your using a different container in your toolchain config (see [this comment](https://github.com/kcoul/SimpleDJ/pull/2#discussion_r1254529644)) for this project. I have so say I did not run into this error when building the app so I would love to maybe follow up with you on comms to debug.

**Pull request merged**